### ISSUE TITLE
Problem of optimisation with Tensorflow

### ISSUE THREAD

**Issue by XiaoyuHy** [tensorflow/tensorflow#49189]
Hello,

My Tensorflow version is as below:

v1.12.1-53126-gd2083b259d1 (tf.version.GIT_VERSION),  2.5.0-dev20210318 (f.version.VERSION).

When I performed training using Gradient Tapes for deep GP models (https://github.com/FelixOpolka/Deep-Gaussian-Process), the code for one iteration of optimisation is as below:

for x_batch, y_batch in zip(bch_X, bch_Y):
	with tf.GradientTape(watch_accessed_variables=False) as tape:
		tape.watch(dgp.trainable_variables)
		objective = -model.elbo((x_batch, y_batch))
		gradients = tape.gradient(objective, model.trainable_variables)
	optimizer.apply_gradients(zip(gradients, model.trainable_variables))
	elbos.append(-objective.numpy())
When I do multiple iterations of optimisations, say 100 iterations, the optimiser may fail and crash at some iteration, with error being a matrix inversion problem. It means the parameters reach a space which would make the matrix non-invertible. However, optimisers if they can compute the objective at the start should not fail and crash when they try parameters that fail the objective (they should just keep the current objective and sample somewhere else, as what the Scipy optimiser does.

Just wondered if there is any solution for this?

Many thanks.

**Comment by saikumarchalla** [comment]
@XiaoyuHy  Please provide the simple standalone code/ colab link along with input data to reproduce the issue at our end.Thanks

**Comment by hosford42** [comment]
@XiaoyuHy You may this function useful. You can pass a tensor through this function to zero out any undefined gradients so they don't reach the trained parameters and corrupt them. The function doesn't affect the value of the tensor, only the gradients applied to it.

```python3
@tf.custom_gradient
def mask_undefined_gradients(x) -> tf.Tensor:
    """Replace undefined values in a tensor's gradient with zeros to 
    avoid corrupting a partially trained model."""

    def grad(upstream):
        return tf.where(tf.math.is_finite(upstream), upstream, 0.0)

    return x, grad
```

If you'd prefer to get a hard error rather than masking the gradients, you can change the `grad` function to assert that none of the upstream gradients are undefined, and then return them unchanged, like so:

```python3
@tf.custom_gradient
def block_undefined_gradients(x) -> tf.Tensor:
    """Raise an exception if there are undefined values in a tensor's 
    gradient to avoid corrupting a partially trained model."""

    def grad(upstream):
        tf.debugging.assert_all_finite(upstream)
        return upstream

    return x, grad
```
